crates/matrix-sdk-crypto/src/store/locks.rs

```diff
@@ -0,0 +1,226 @@
+    const MAX_BACKOFF_MS: u32 = 1000;
+
+    /// Create a new store-based lock implemented as a value in the
+    /// crypto-store.
+    ///
+    /// # Parameters
+    ///
+    /// - `lock_key`: key in the key-value store to store the lock's state.
+    /// - `lock_holder`: identify the lock's holder with this given value.
+    /// - `max_backoff`: maximum time (in milliseconds) that should be waited
+    ///   for, between two
+    /// attempts. When that time is reached a second time, the lock will stop
+    /// attempting to get the lock and will return a timeout error upon
+    /// locking. If not provided, will wait for [`Self::MAX_BACKOFF_MS`].
+    pub fn new(
+        store: Arc<DynCryptoStore>,
+        lock_key: String,
+        lock_holder: String,
+        max_backoff: Option<u32>,
+    ) -> Self {
+        let max_backoff = max_backoff.unwrap_or(Self::MAX_BACKOFF_MS);
+        Self { store, lock_key, lock_holder, max_backoff, backoff: Self::INITIAL_BACKOFF_MS }
+    }
+
+    /// Attempt to take the lock, with exponential backoff if the lock has
+    /// already been taken before.
+    pub async fn lock(&mut self) -> Result<(), CryptoStoreError> {
+        loop {
+            let inserted = self
+                .store
+                .insert_custom_value_if_missing(
+                    &self.lock_key,
+                    self.lock_holder.as_bytes().to_vec(),
+                )
+                .await?;
+
+            if inserted {
+                // Reset backoff before returning, for the next attempt to lock.
+                self.backoff = Self::INITIAL_BACKOFF_MS;
+                return Ok(());
+            }
+
+            // Double-check that we were not interrupted last time we tried to take the
+            // lock, and forgot to release it; in that case, we *still* hold it.
+            let previous = self.store.get_custom_value(&self.lock_key).await?;
+            if previous.as_deref() == Some(self.lock_holder.as_bytes()) {
+                // At this point, the only possible value for backoff is the initial one, but
+                // better be safe than sorry.
+                tracing::warn!(
+                    "Crypto-store lock {} was already taken by {}; let's pretend we just acquired it.",
+                    self.lock_key,
+                    self.lock_holder
+                );
+                self.backoff = Self::INITIAL_BACKOFF_MS;
+                return Ok(());
+            }
+
+            // Exponential backoff! Multiply by 2 the time we've waited before, cap it to
+            // max_backoff.
+            let wait = self.backoff;
+
+            if wait == u32::max_value() {
+                // We've reached the maximum backoff, abandon.
+                return Err(LockStoreError::LockTimeout.into());
+            }
+
+            self.backoff = self.backoff.saturating_mul(2);
+            if self.backoff >= self.max_backoff {
+                self.backoff = u32::max_value();
+            }
```

Comment:
Hmm yeah, semantics are slightly different, since the sqlite-based lock will also timeout.

But at this point, I'm realizing timing out might not even be necessary for the #1928 use case?

- The NSE process would either try to take the lock, and if it's already taken, it would probably not even wait and continue.
- The main process would try to take the lock until the NSE process was done with it.

Reply:
Yep, and on the second point, the main process will want to hold on to the lock as long as it's alive, not sure if that would be problematic for releasing the lock if we get killed.